R/block_cv.R

```diff
@@ -0,0 +1,172 @@
+        i = "Try converting `data` to an `sf` object via `sf::st_as_sf()`."
+      )
+    )
+  }
+
+  if (sf::st_crs(data) == sf::NA_crs_) {
+    rlang::abort(
+      c(
+        "`spatial_block_cv()` requires your data to have an appropriate coordinate reference system (CRS).",
+        i = "Try setting a CRS using `sf::st_set_crs()`."
+      )
+    )
+  }
+
+  grid_box <- sf::st_bbox(data)
+  if (sf::st_is_longlat(data)) {
+    # cf https://github.com/ropensci/stplanr/pull/467
+    # basically: spherical geometry means sometimes the straight line of the
+    # grid will exclude points within the bounding box
+    #
+    # so here we'll expand our boundary by 0.1% in order to always contain our
+    # points within the grid
+    grid_box[1] <- grid_box[1] - abs(grid_box[1] * 0.001)
+    grid_box[2] <- grid_box[2] - abs(grid_box[2] * 0.001)
+    grid_box[3] <- grid_box[3] + abs(grid_box[3] * 0.001)
+    grid_box[4] <- grid_box[4] + abs(grid_box[4] * 0.001)
+  }
+  grid_blocks <- sf::st_make_grid(grid_box, ...)
+  split_objs <- switch(
+    method,
+    "random" = random_block_cv(data, grid_blocks, v)
+  )
+  v <- split_objs$v[[1]]
+  split_objs$v <- NULL
+
+  ## We remove the holdout indices since it will save space and we can
+  ## derive them later when they are needed.
+  split_objs$splits <- map(split_objs$splits, rm_out)
+
+  ## Save some overall information
+  cv_att <- list(v = v)
+
+  new_rset(
+    splits = split_objs$splits,
+    ids = split_objs[, grepl("^id", names(split_objs))],
+    attrib = cv_att,
+    subclass = c("spatial_block_cv", "rset")
+  )
+
+}
+
+random_block_cv <- function(data, grid_blocks, v) {
+  n <- nrow(data)
+
+  block_contains_points <- purrr::map_lgl(
+    sf::st_intersects(grid_blocks, data),
+    sgbp_is_not_empty
+  )
+  grid_blocks <- grid_blocks[block_contains_points]
+
+  n_blocks <- length(grid_blocks)
+  if (!is.numeric(v) || length(v) != 1) {
+    rlang::abort("`v` must be a single integer.")
+  }
+  if (v > n_blocks) {
+    rlang::warn(paste0(
+      "Fewer than ", v, " blocks available for sampling; setting v to ",
+      n_blocks, "."
+    ))
+    v <- n_blocks
+  }
+
+  grid_blocks <- sf::st_as_sf(grid_blocks)
+  grid_blocks$fold <- sample(rep(seq_len(v), length.out = nrow(grid_blocks)))
+  grid_blocks <- split_unnamed(grid_blocks, grid_blocks$fold)
+
+  # grid_blocks is now a list of sgbp lists (?sf::sgbp)
+  #
+  # The first map() here iterates through the meta-list,
+  # and the second checks each element of the relevant sgbp list
+  # to see if it is integer(0) (no intersections) or not
+  #
+  # Each sgbp sub-list is nrow(data) elements long, so this which()
+  # returns the list indices which are not empty, which is equivalent
+  # to the row numbers that intersect with blocks in the fold
+  indices <- purrr::map(
+    grid_blocks,
+    function(blocks) which(
+      purrr::map_lgl(
+        sf::st_intersects(data, blocks),
```

Comment:
@hfrick I added a comment above here and also made the arguments a little more explicit, so now it's a nested map() rather than a vapply() inside the map(). Does this parse better? I think I prefer seeing the st_intersects explicitly rather than a helper.

Reply:
I like this better 👍

Reply:
yeah, I agree, it's good to see the order of the args going into `st_intersects()`. this snippet now definitely purrrs 😸